src/utils/common.ts

```diff
@@ -267,11 +269,20 @@ export async function estimateTotalGasRequiredByUnsignedTransaction(
   if (chainIsOPStack(chainId)) {
     assert(isOptimismL2Provider(provider), `Unexpected provider for chain ID ${chainId}.`);
     assert(gasPrice === undefined, `Gas price (${gasPrice}) supplied for Optimism gas estimation (unused).`);
-    const populatedTransaction = await voidSigner.populateTransaction(unsignedTx);
-    tokenGasCost = await provider.estimateTotalGasCost(populatedTransaction);
+    const populatedTransaction = await voidSigner.populateTransaction({
+      ...unsignedTx,
+      gasLimit: nativeGasCost, // prevents additional gas estimation call
+    });
+    // Concurrently estimate the gas cost on L1 and L2 instead of calling
+    // `provider.estimateTotalGasCost` to improve performance.
+    const [l1GasCost, l2GasCost] = await Promise.all([
+      provider.estimateL1GasCost(populatedTransaction),
+      provider.estimateL2GasCost(populatedTransaction),
```

Comment:
Where are these functions defined?

Reply:
These are from the Optimism SDK https://sdk.optimism.io/modules#estimateL2GasCost. But can't find the source code on GitHub because the SDK is superseded now by https://viem.sh/op-stack

Reply:
These are OP-specific implementations (note `if (chainIsOPStack(chainId))` a few lines above)- see https://sdk.optimism.io/modules#estimateL1GasCost

Reply:
Thank u both!

Reply:
@dohaki Running these in paralell is a nice observation based on the implementation of `estimateTotalGasCost()` here: https://github.com/ethereum-optimism/optimism/blob/ee058e413e0c7ff92d000593af648f3d79ca1a31/packages/sdk/src/l2-provider.ts#L133